### PR TITLE
Disable custom ad notifications in release and beta builds

### PR DESCRIPTION
**Pull request by tmancey**
Resolves https://github.com/brave/brave-browser/issues/13363

## Submitter Checklist:

- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed).
- [ ] Requested a security/privacy review as needed.

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

Custom ad notifications should only be shown on nightly builds. Release and beta channels should show native ad notifications

### CODE REVIEW

File: android/BUILD.gn [hunk 1]
```diff
@@ -6,6 +6,12 @@ declare_args() {
   brave_android_developer_options_code = ""
 }
 
+java_cpp_enum("brave_android_java_enums_srcjar") {
```
Comment (tmancey) [inline android/BUILD.gn:9]: @yachtcaptain23 Do you remember why this was added back as this was not part of the original ad notifications?

**Comment by tmancey** [comment]
@yachtcaptain23 Do we not need to add `//brave/browser/notifications` back to `browser/BUILD.gn` or does the `brave_ads` implementation work for both?

**Comment by tmancey** [comment]
@yachtcaptain23 Should `@JNINamespace("brave_ads")` be reverted back to `@JNINamespace("chrome::android")`?

**Comment by tmancey** [comment]
@yachtcaptain23 `browser/notifications/brave_notification_platform_bridge_helper_android.cc` and `browser/notifications/brave_notification_platform_bridge_helper_android.h` changes for native ad notifications have not been reverted

**Comment by tmancey** [comment]
@yachtcaptain23 was it intentional that `components/brave_ads/browser/ads_notification_handler.cc` and `components/brave_ads/browser/ads_notification_handler.h` changes for native ad notifications have not been reverted?

File: components/brave_ads/browser/ads_service_impl.cc [hunk 2]
```diff
@@ -81,6 +83,7 @@
 #include "third_party/dom_distiller_js/dom_distiller.pb.h"
 #include "ui/base/l10n/l10n_util.h"
 #include "ui/base/resource/resource_bundle.h"
+#include "ui/message_center/public/cpp/notification.h"
 
 #if defined(OS_ANDROID)
 #include "chrome/browser/android/service_tab_launcher.h"
```
Comment (tmancey) [inline components/brave_ads/browser/ads_service_impl.cc:89]: @yachtcaptain23 Is `#include "brave/browser/notifications/brave_notification_platform_bridge_helper_android.h"` not needed here, as it was included before we moved to custom ad notifications?

File: components/brave_ads/browser/ads_service_impl.cc [hunk 3]
```diff
@@ -1709,14 +1740,22 @@ std::string AdsServiceImpl::LoadDataResourceAndDecompressIfNeeded(
...
 
-  platform_bridge->Display(profile_, notification);
+    platform_bridge->Display(profile_, notification);
+  } else {
+    auto notification = CreateMessageCenterNotification(ad_notification);
+    display_service_->Display(NotificationHandler::Type::BRAVE_ADS,
+        *notification, /*metadata=*/nullptr);
+  }
 
   StartNotificationTimeoutTimer(ad_notification.uuid);
```
Comment (tmancey) [inline components/brave_ads/browser/ads_service_impl.cc:1760]: @yachtcaptain23 `StartNotificationTimeoutTimer` was not previously called for Android for native notifications, is this logic correct?

File: components/brave_ads/browser/ads_service_impl.cc [hunk 4]
```diff
@@ -1759,9 +1798,13 @@ bool AdsServiceImpl::ShouldShowNotifications() {
...
     const std::string& uuid) {
-  std::unique_ptr<PlatformBridge>
-    platform_bridge = std::make_unique<PlatformBridge>(profile_);
-  platform_bridge->Close(profile_, uuid);
+  if (brave::IsDevOrCanaryBuild()) {
+    std::unique_ptr<PlatformBridge>
+      platform_bridge = std::make_unique<PlatformBridge>(profile_);
+    platform_bridge->Close(profile_, uuid);
+  } else {
+    display_service_->Close(NotificationHandler::Type::BRAVE_ADS, uuid);
```
Comment (tmancey) [inline components/brave_ads/browser/ads_service_impl.cc:1806]: Not all changes have been reverted, see https://github.com/brave/brave-core/pull/6728

File: components/brave_ads/browser/ads_service_impl.cc [hunk 5]
```diff
@@ -658,6 +662,10 @@ void AdsServiceImpl::OnInitialize(
 
   is_initialized_ = true;
 
+  if (!brave::IsDevOrCanaryBuild()) {
+    SetAdsServiceForNotificationHandler();
```
Comment (tmancey) [inline components/brave_ads/browser/ads_service_impl.cc:666]: @yachtcaptain23 Ads service should only be set for native ad notifications, i.e. non nightly builds. Can you please code review the `!brave::IsDevOrCanaryBuild()` guard I have added

Comment (yachtcaptain23) [reply components/brave_ads/browser/ads_service_impl.cc:666]: LG2M

File: components/brave_ads/browser/ads_service_impl.cc (hunk 4 above)
Comment (yachtcaptain23) [reply components/brave_ads/browser/ads_service_impl.cc:1806]: I will need to dive a bit deeper on this, will take a look.

File: components/brave_ads/browser/ads_service_impl.cc (hunk 2 above)
Comment (yachtcaptain23) [reply components/brave_ads/browser/ads_service_impl.cc:89]: Will follow up

Comment (yachtcaptain23) [reply components/brave_ads/browser/ads_service_impl.cc:89]: I'll be testing and posting the results, but the code defines `MaybeRegenerateNotification` which we don't use.

File: chromium_src/chrome/browser/notifications/notification_handler.h [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    WEB_PERSISTENT = 0,
+    WEB_NON_PERSISTENT = 1,
+    EXTENSION = 2,
+    SEND_TAB_TO_SELF = 3,
+    TRANSIENT = 4,
+    PERMISSION_REQUEST = 5,
+    SHARING = 6,
+    ANNOUNCEMENT = 7,
+    NEARBY_SHARE = 8,
+    BRAVE_ADS = 9,
```
Comment (yachtcaptain23) [inline chromium_src/chrome/browser/notifications/notification_handler.h:27]: Would you apply the following?

```
diff --git a/chromium_src/chrome/browser/notifications/notification_handler.h b/chromium_src/chrome/browser/notifications/notification_handler.h
index 761c28e67b..2941684004 100644
--- a/chromium_src/chrome/browser/notifications/notification_handler.h
+++ b/chromium_src/chrome/browser/notifications/notification_handler.h
@@ -24,7 +24,8 @@ class NotificationHandler : public NotificationHandler_ChromiumImpl {
     SHARING = 6,
     ANNOUNCEMENT = 7,
     NEARBY_SHARE = 8,
-    BRAVE_ADS = 9,
+    NOTIFICATIONS_MUTED = 9,
+    BRAVE_ADS = 10,
     MAX = BRAVE_ADS,
   };
```
Chromium team added `NOTIFICATIONS_MUTED` in October

Comment (tmancey) [reply chromium_src/chrome/browser/notifications/notification_handler.h:27]: Yes please

File: components/brave_ads/browser/ads_service_impl.cc (hunk 3 above)
Comment (tmancey) [reply components/brave_ads/browser/ads_service_impl.cc:1760]: Discussed with @yachtcaptain23 and resolved by not starting the timeout timer for release or beta builds on Android

File: components/brave_ads/browser/ads_service_impl.cc (hunk 4 above)
Comment (tmancey) [reply components/brave_ads/browser/ads_service_impl.cc:1812]: Agreed with @yachtcaptain23 this logic is required to limit the amount of ad notifications that appear in the system tray on Android

File: components/brave_ads/browser/ads_service_impl.cc (hunk 2 above)
Comment (tmancey) [reply components/brave_ads/browser/ads_service_impl.cc:89]: Adding back as required to limit amount of ad notifications in the system tray on Android

**Comment by tmancey** [comment]
@yachtcaptain23 Can you please sanity test on Android and Linux.

**Android (native):**
View 4 ads and confirm they appear in the system tray
Confirm background ad notifications are shown before 9pm and after 6am
Confirm background ad notifications are not show between 9pm and 6am
Confirm ads can be viewed
Confirm ads can be clicked
Confirm ads can be dismissed
Confirm ads timeout

**Android (custom):**
Confirm ads can be viewed
Confirm ads can be clicked
Confirm ads can be dismissed
Confirm ads timeout

**Linux (native/custom):**
Confirm ads can be viewed
Confirm ads can be clicked
Confirm ads can be dismissed
Confirm ads timeout

**Comment by tmancey** [comment]
@yachtcaptain23 as discussed `components/brave_ads/browser/ads_notification_handler.cc` and `components/brave_ads/browser/ads_notification_handler.h` changes need reverting for native ad notifications

**Review by kylehickinson — APPROVED** [review]
iOS lgtm++

**Review by samartnik — APPROVED** [review]
Java patch looks good

File: chromium_src/chrome/browser/notifications/notification_handler.h [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+ public:
+  // GENERATED_JAVA_ENUM_PACKAGE: org.chromium.chrome.browser.notifications
+  // GENERATED_JAVA_CLASS_NAME_OVERRIDE: NotificationType
+  enum class Type {
+    WEB_PERSISTENT = 0,
+    WEB_NON_PERSISTENT = 1,
+    EXTENSION = 2,
+    SEND_TAB_TO_SELF = 3,
+    TRANSIENT = 4,
+    PERMISSION_REQUEST = 5,
```
Comment (mkarolin) [inline chromium_src/chrome/browser/notifications/notification_handler.h:23]: This value has been deprecated upstream.

File: chromium_src/chrome/browser/notifications/notification_handler.h [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+    SHARING = 6,
+    ANNOUNCEMENT = 7,
+    NEARBY_SHARE = 8,
+    NOTIFICATIONS_MUTED = 9,
+    BRAVE_ADS = 10,
```
Comment (mkarolin) [inline chromium_src/chrome/browser/notifications/notification_handler.h:28]: In 1.19.x branch we have this value as 9. @iefremov was suggesting here https://github.com/brave/brave-core/pull/6957#discussion_r526018605 that perhaps we should keep the same value and let upstream grow above. For cr88 uplift to 1.19.x I was planning to do this: https://github.com/brave/brave-core/commit/f3fcc4192276b5ad9192c9482849053c7b1ae6e9.

File: chromium_src/chrome/browser/notifications/notification_platform_bridge_mac.mm [hunk 9]
```diff
@@ -0,0 +1,79 @@
+#include "chrome/browser/notifications/notification_platform_bridge_mac.h"
```
Comment (mkarolin) [inline chromium_src/chrome/browser/notifications/notification_platform_bridge_mac.mm:1]: Copyright statements are missing.

File: chromium_src/chrome/browser/notifications/notification_platform_bridge_mac.mm [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+  [timers_ removeObject:timer];
+}
+
+@end
+
+static NotificationTimeoutMac*
+    g_notification_platform_bridge_notification_timeout =
+        [[[NotificationTimeoutMac alloc] init] retain];
+
+#define BRAVE_DISPLAY_ \
```
Comment (mkarolin) [inline chromium_src/chrome/browser/notifications/notification_platform_bridge_mac.mm:71]: This doesn't seem to be used. Is there a patch missing?

File: chromium_src/chrome/browser/notifications/notification_platform_bridge_mac.mm [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+  [g_notification_platform_bridge_notification_timeout startTimer:toast];
+
+#define BRAVE_SUPPORTSALERTSIMPL_ \
```
Comment (mkarolin) [inline chromium_src/chrome/browser/notifications/notification_platform_bridge_mac.mm:74]: Ditto.

File: chromium_src/chrome/browser/notifications/notification_platform_bridge_mac.mm (hunk 10 above)
Comment (tmancey) [reply chromium_src/chrome/browser/notifications/notification_platform_bridge_mac.mm:71]: @yachtcaptain23 ^^^

File: chromium_src/chrome/browser/notifications/notification_platform_bridge_mac.mm (hunk 11 above)
Comment (tmancey) [reply chromium_src/chrome/browser/notifications/notification_platform_bridge_mac.mm:74]: @yachtcaptain23 ^^^

File: patches/chrome-browser-notifications-notification_display_service_impl.cc.patch [hunk 12]
```diff
@@ -0,0 +1,12 @@
+diff --git a/chrome/browser/notifications/notification_display_service_impl.cc b/chrome/browser/notifications/notification_display_service_impl.cc
+index 286b6af7d35c5acdebcbb41bc9acf468f93f5f0d..f168680bee4cbc64519004c38f92f2d80a2cb482 100644
+--- a/chrome/browser/notifications/notification_display_service_impl.cc
++++ b/chrome/browser/notifications/notification_display_service_impl.cc
+@@ -100,6 +100,7 @@ NotificationDisplayServiceImpl::NotificationDisplayServiceImpl(Profile* profile)
+                              std::make_unique<NearbyNotificationHandler>());
+     }
+ #endif
++    BRAVE_ADD_BRAVE_ADS_NOTIFICATION_HANDLER;
```
Comment (mkarolin) [inline patches/chrome-browser-notifications-notification_display_service_impl.cc.patch:9]: Would subclassing be possible instead?

File: chromium_src/chrome/browser/notifications/notification_handler.h (hunk 8 above)
Comment (tmancey) [reply chromium_src/chrome/browser/notifications/notification_handler.h:28]: @mkarolin I will revert to 9 and remove `NOTIFICATIONS_MUTED`, thanks

File: chromium_src/chrome/browser/notifications/notification_handler.h (hunk 7 above)
Comment (tmancey) [reply chromium_src/chrome/browser/notifications/notification_handler.h:23]: If works with you will leave as is until patched?

File: chromium_src/chrome/browser/notifications/notification_handler.h (hunk 8 above)
Comment (tmancey) [reply chromium_src/chrome/browser/notifications/notification_handler.h:28]: @mkarolin Fixed

File: chromium_src/chrome/browser/notifications/notification_platform_bridge_mac.mm (hunk 9 above)
Comment (tmancey) [reply chromium_src/chrome/browser/notifications/notification_platform_bridge_mac.mm:1]: Fixed (this is just a revert of a previous PR as we now need both native and custom ad notifications)

File: patches/chrome-browser-notifications-notification_platform_bridge_android.h.patch [hunk 13]
```diff
@@ -0,0 +1,12 @@
+diff --git a/chrome/browser/notifications/notification_platform_bridge_android.h b/chrome/browser/notifications/notification_platform_bridge_android.h
+index 7ff354723b9d14c43b9dd7dd2679f36ab8e66f79..ffb5f27cbc4f7d5f826e4717c9a6d6914abb1949 100644
+--- a/chrome/browser/notifications/notification_platform_bridge_android.h
++++ b/chrome/browser/notifications/notification_platform_bridge_android.h
+@@ -88,6 +88,7 @@ class NotificationPlatformBridgeAndroid : public NotificationPlatformBridge {
+   static void RegisterProfilePrefs(user_prefs::PrefRegistrySyncable* registry);
+
+  private:
++  friend class BraveNotificationPlatformBridgeHelperAndroid;
```
Comment (bridiver) [inline patches/chrome-browser-notifications-notification_platform_bridge_android.h.patch:9]: always use defines to make patches extensible

File: chromium_src/chrome/browser/notifications/notification_platform_bridge_mac.mm (hunk 10 above)
Comment (tmancey) [reply chromium_src/chrome/browser/notifications/notification_platform_bridge_mac.mm:76]: @mkarolin Fixed, thanks

File: chromium_src/chrome/browser/notifications/notification_platform_bridge_mac.mm (hunk 11 above)
Comment (tmancey) [reply chromium_src/chrome/browser/notifications/notification_platform_bridge_mac.mm:79]: @mkarolin Fixed, thanks

File: patches/chrome-browser-notifications-notification_display_service_impl.cc.patch (hunk 12 above)
Comment (tmancey) [reply patches/chrome-browser-notifications-notification_display_service_impl.cc.patch:9]: This is only temporary until we move to custom ad notifications across all build channels. This PR is reverting previous changes so would rather not change any logic. Thanks

File: patches/chrome-browser-notifications-notification_platform_bridge_android.h.patch (hunk 13 above)
Comment (tmancey) [reply patches/chrome-browser-notifications-notification_platform_bridge_android.h.patch:9]: @yachtcaptain23 ^^^

Comment (tmancey) [reply patches/chrome-browser-notifications-notification_platform_bridge_android.h.patch:9]: This is only temporary until we move to custom ad notifications across all build channels. This PR is reverting previous changes so would rather not change any logic. Thanks

File: chromium_src/chrome/browser/notifications/notification_handler.h (hunk 6 above)
Comment (mkarolin) [reply chromium_src/chrome/browser/notifications/notification_handler.h:27]: Pretty sure you'll get build errors without NOTIFICATIONS_MUTED in the enum.

File: chromium_src/chrome/browser/notifications/notification_handler.h (hunk 8 above)
Comment (mkarolin) [reply chromium_src/chrome/browser/notifications/notification_handler.h]: I don't think we can remove `NOTIFICATIONS_MUTED` - we'd get build errors. The suggestion was to leave `BRAVE_ADS = 9` and make `NOTIFICATIONS_MUTED = 10` (and update the DCHECK).

File: chromium_src/chrome/browser/notifications/notification_handler.h (hunk 7 above)
Comment (mkarolin) [reply chromium_src/chrome/browser/notifications/notification_handler.h:23]: Not sure I am following. cr88 is already in `master` and it's already deprecated in cr88.

File: chromium_src/chrome/browser/notifications/notification_handler.h (hunk 6 above)
Comment (tmancey) [reply chromium_src/chrome/browser/notifications/notification_handler.h:27]: Fixed, thanks

File: chromium_src/chrome/browser/notifications/notification_handler.h (hunk 7 above)
Comment (tmancey) [reply chromium_src/chrome/browser/notifications/notification_handler.h:23]: @mkarolin apologies, misunderstood. Fixed

File: chromium_src/chrome/browser/notifications/notification_handler.h (hunk 8 above)
Comment (tmancey) [reply chromium_src/chrome/browser/notifications/notification_handler.h]: Fixed

**Review by mkarolin — APPROVED** [review]
++ for `chromium_src` and `patches`

**Comment by tmancey** [comment]
> @yachtcaptain23 as discussed `components/brave_ads/browser/ads_notification_handler.cc` and `components/brave_ads/browser/ads_notification_handler.h` changes need reverting for native ad notifications

@yachtcaptain23 Confirmed "so the headless check was basically a timer to open and close the notification, we don't need it. I think when I had worked on custom notifications, I had rewritten it so there wouldn’t be a need for this headless timer"

File: vendor/bat-native-ads/BUILD.gn [hunk 14]
```diff
@@ -573,6 +573,7 @@ source_set("ads") {
 
   deps = [
     "//base",
+    "//brave/common",
```
Comment (bridiver) [inline vendor/bat-native-ads/BUILD.gn:576]: This is a layering violation and is not allowed here, nor is it allowed in components/brave_ads

File: chromium_src/chrome/browser/notifications/notification_handler.h [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+#define NotificationHandler NotificationHandler_ChromiumImpl
+#include "../../../../../chrome/browser/notifications/notification_handler.h"
+#undef NotificationHandler
+
+class NotificationHandler : public NotificationHandler_ChromiumImpl {
+ public:
+  // GENERATED_JAVA_ENUM_PACKAGE: org.chromium.chrome.browser.notifications
+  // GENERATED_JAVA_CLASS_NAME_OVERRIDE: NotificationType
+  enum class Type {
+    WEB_PERSISTENT = 0,
```
Comment (bridiver) [inline chromium_src/chrome/browser/notifications/notification_handler.h:18]: this duplication of upstream is not ok

Comment (bridiver) [reply chromium_src/chrome/browser/notifications/notification_handler.h:18]: this should be `#define MAX ... brave stuff goes here... MAX`

File: chromium_src/chrome/browser/notifications/notification_handler.h (hunk 6 above)
Comment (bridiver) [reply chromium_src/chrome/browser/notifications/notification_handler.h:27]: why are we adding BRAVE_ADS in the middle instead of the end?

**Pull request merged**